### PR TITLE
[locator] active filter lists all matching fields as results when typing '@'

### PR DESCRIPTION
**Pull request by 3nids**
results can be triggered to restrict the search to the given field

image

### CODE REVIEW

File: tests/src/app/testqgsapplocatorfilters.cpp [hunk 1]
```diff
@@ -220,6 +220,16 @@ void TestQgsAppLocatorFilters::testActiveLayerFieldRestriction()
   restr = QgsActiveLayerFeaturesLocatorFilter::fieldRestriction( search );
   QCOMPARE( restr, QStringLiteral( "home" ) );
   QCOMPARE( search, QStringLiteral( "" ) );
+
+  search = QStringLiteral( "@" );
+  restr = QgsActiveLayerFeaturesLocatorFilter::fieldRestriction( search );
+  QVERIFY( !restr.isNull() );
+  QCOMPARE( search, QStringLiteral( "" ) );
```
Comment (nyalldawson) [inline tests/src/app/testqgsapplocatorfilters.cpp:227]:
```suggestion
  QCOMPARE( search, QString() );
```

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 2]
```diff
@@ -361,6 +372,17 @@ void QgsActiveLayerFeaturesLocatorFilter::fetchResults( const QString &string, c
   QString searchString = string;
   fieldRestriction( searchString );
 
+  // propose available fields for restriction
+  for ( const QString &field : qgis::as_const( mFieldsCompletion ) )
+  {
+    QgsLocatorResult result;
+    result.displayString = QStringLiteral( "@%1" ).arg( field );
+    result.description = QStringLiteral( "Limit the search to the field '%1'" ).arg( field );
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:380]: tr

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 3]
```diff
@@ -258,8 +259,10 @@ QString QgsActiveLayerFeaturesLocatorFilter::fieldRestriction( QString &searchSt
 
 QStringList QgsActiveLayerFeaturesLocatorFilter::prepare( const QString &string, const QgsLocatorContext &context )
 {
-  // Normally skip very short search strings, unless when specifically searching using this filter
-  if ( string.length() < 3 && !context.usingPrefix )
+  mFieldsCompletion.clear();
+
+  // Normally skip very short search strings, unless when specifically searching using this filter or try to match fields
+  if ( string.length() < 3 && !context.usingPrefix && !string.startsWith( QStringLiteral( "@" ) ) )
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:265]:
```suggestion
  if ( string.length() < 3 && !context.usingPrefix && !string.startsWith( '@' ) ) )
```

File: src/app/locator/qgsinbuiltlocatorfilters.cpp [hunk 4]
```diff
@@ -308,16 +311,24 @@ QStringList QgsActiveLayerFeaturesLocatorFilter::prepare( const QString &string,
     if ( field.configurationFlags().testFlag( QgsField::ConfigurationFlag::NotSearchable ) )
       continue;
 
-    if ( !_fieldRestriction.isEmpty() && !field.name().startsWith( _fieldRestriction ) )
+    if ( !_fieldRestriction.isNull() && !field.name().startsWith( _fieldRestriction ) )
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.cpp:314]: why this change? it's a step back, in that the concept of null QStrings is deprecated...

File: src/app/locator/qgsinbuiltlocatorfilters.h [hunk 5]
```diff
@@ -123,11 +141,15 @@ class APP_EXPORT QgsActiveLayerFeaturesLocatorFilter : public QgsLocatorFilter
     QString mLayerId;
     QIcon mLayerIcon;
     QStringList mAttributeAliases;
+    QStringList mFieldsCompletion;
     int mMaxTotalResults = 30;
 
     friend class TestQgsAppLocatorFilters;
 };
 
+Q_DECLARE_METATYPE( QgsActiveLayerFeaturesLocatorFilter::Data )
```
Comment (nyalldawson) [inline src/app/locator/qgsinbuiltlocatorfilters.h:150]: Would be cleaner to avoid the Data struct and instead use a QVariantMap value.

File: src/app/locator/qgsinbuiltlocatorfilters.cpp (hunk 4 above)
Comment (3nids) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:314]: I wanted to distinguish when '@' was typed with nothing yet. But I can replace it with a `startsWith('@')`.

Comment (nyalldawson) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp:314]: Why doesn't testing .isEmpty() work for the same result?

Comment (3nids) [reply src/app/locator/qgsinbuiltlocatorfilters.cpp]: My first idea was:
search: 'something' => fieldRestriction is null
search: '@field something' => fieldRestriction is field
search: '@' => fieldRestriction is empty string, not null

I added now a pointer to a boolean in the private method, things are a bit more explicit like this.

File: src/app/locator/qgsinbuiltlocatorfilters.h (hunk 5 above)
Comment (3nids) [reply src/app/locator/qgsinbuiltlocatorfilters.h]: done
